search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,568 @@
+                         GDBusMethodInvocation     *invocation,
+                         gchar                    **results,
+                         gpointer                   user_data)
+{
+
+    DelayedInitialResultSet *delayed;
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+
+    if (g_hash_table_size (app->page_data_hash_map) > 0) {
+        g_dbus_method_invocation_return_value (invocation,
+                                               get_result_metas (results,
+                                                                 app->page_data_hash_map));
+        return TRUE;
+    }
+
+    delayed = delayed_initial_result_set_new (app,
+                                              invocation,
+                                              results,
+                                              get_result_metas);
+    g_idle_add_full (G_PRIORITY_DEFAULT_IDLE,
+                     check_page_data,
+                     delayed,
+                     (GDestroyNotify) delayed_initial_result_set_free);
+    return TRUE;
+}
+
+static gboolean
+yelp_launch (gchar                 *terms,
+             gpointer               user_data,
+             GDBusMethodInvocation *invocation,
+             gchar                 *action)
+{
+    GDBusConnection *session_bus;
+    GVariant *reply;
+    GError *error = NULL;
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GVariantBuilder param_builder;
+    gint i;
+
+    session_bus = g_dbus_interface_skeleton_get_connection (G_DBUS_INTERFACE_SKELETON (app->skeleton));
+    if (session_bus == NULL) {
+        g_dbus_method_invocation_return_error_literal (invocation,
+                                                       G_IO_ERROR,
+                                                       G_IO_ERROR_FAILED,
+                                                       "failed to get connection to the session bus");
+        return TRUE;
+    }
+
+    g_variant_builder_init (&param_builder, G_VARIANT_TYPE ("av"));
+    g_variant_builder_add (&param_builder, "v", g_variant_new_string (terms));
+
+    reply = g_dbus_connection_call_sync (session_bus,
+                                         "org.gnome.Yelp",
+                                         "/org/gnome/Yelp",
+                                         "org.freedesktop.Application",
+                                         "ActivateAction",
+                                         g_variant_new ("(sava{sv})",
+                                                        action,
+                                                        &param_builder,
+                                                        NULL),
+                                         NULL,
+                                         G_DBUS_CALL_FLAGS_NONE,
+                                         -1,
+                                         NULL,
+                                         &error);
+
+    g_object_unref (session_bus);
```

Comment:
This should not be unreffed - `g_dbus_interface_skeleton_get_connection` does not add a reference to the returned `GDBusConnection` instance (it has `transfer none` annotation).